File: Leetcode/105.cpp
/**
 * 105. 从前序与中序遍历序列构造二叉树  迭代法
 * link：https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
 */

class Solution
{
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        if (preorder.empty() || inorder.empty())
            return nullptr;

        int n = preorder.size();
        int preIndex = 0;
        int inIndex = 0;
        TreeNode *root = new TreeNode(preorder[preIndex++]);
        stack<TreeNode *> rootNodes({root});

        while (preIndex < n)
        {
            TreeNode *curRoot = nullptr;

            while (!rootNodes.empty() && inorder[inIndex] == rootNodes.top()->val)
            {
                curRoot = rootNodes.top();
                rootNodes.pop();
                inIndex++;
            }

            TreeNode *cur = new TreeNode(preorder[preIndex++]);
            if (curRoot != nullptr)
            {
                curRoot->right = cur;
            }
            else
            {
                rootNodes.top()->left = cur;
            }

            rootNodes.push(cur);
        }
        return root;
    }
};

File: Leetcode/105.py
# 105. 从前序与中序遍历序列构造二叉树  迭代法
# link: https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder or not inorder:
            return None

        pre_idx: int = 0
        in_indx: int = 0
        root: TreeNode = TreeNode(preorder[pre_idx])
        root_nodes = [root]
        pre_idx += 1

        while pre_idx < len(preorder):
            cur_root: TreeNode = None

            while root_nodes and inorder[in_indx] == root_nodes[-1].val:
                cur_root = root_nodes[-1]
                root_nodes.pop()
                in_indx += 1

            cur: TreeNode = TreeNode(preorder[pre_idx])
            pre_idx += 1

            if cur_root:
                cur_root.right = cur
            else:
                root_nodes[-1].left = cur

            root_nodes.append(cur)

        return root
